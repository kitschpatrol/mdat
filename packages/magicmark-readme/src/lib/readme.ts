import rulePresets from './rules'
import { findUp } from 'find-up'
import {
	type CheckFileReport,
	type CheckOptions,
	type ExpandFileOptions,
	type ExpandFileReport,
	checkFile,
	expandFile,
	log,
} from 'magicmark'
import fs from 'node:fs/promises'
import path from 'node:path'
import { packageDirectory } from 'pkg-dir'

export async function checkReadmeFile(
	path: string,
	options: CheckOptions,
): Promise<CheckFileReport> {
	const { meta = true, prefix, rules = [] } = options

	const report = await checkFile(path, {
		meta,
		prefix,
		rules: [rulePresets.readme, ...rules],
	})
	return report
}

export async function expandReadmeFile(
	path: string,
	options: ExpandFileOptions,
): Promise<ExpandFileReport> {
	const { meta = true, name, output, prefix, print, rules = [] } = options

	const report = await expandFile(path, {
		meta,
		name,
		output,
		prefix,
		print,
		rules: [rulePresets.readme, ...rules],
	})
	return report
}

export async function getReadmePath(): Promise<string> {
	log.info(`Searching for package directory...`)

	// First, check for a readme in the package directory
	const thePackageDirectory = await packageDirectory()
	if (thePackageDirectory !== undefined) {
		log.info(`Found package directory: "${thePackageDirectory}"`)
		const files = await fs.readdir(thePackageDirectory)
		const readmeFile = files.find((file) => file.toLowerCase() === 'readme.md')
		if (readmeFile !== undefined) {
			return path.join(thePackageDirectory, readmeFile)
		}

		log.warn(`Readme not found in package directory, searching up for readme...`)
	}

	// FindUp is case-insensitive
	const closestReadme = await findUp('readme.md', {
		type: 'file',
	})

	if (closestReadme !== undefined) {
		log.info(`Found closest readme at "${closestReadme}"`)
		return closestReadme
	}

	throw new Error('No readme found')
}

// Decadent
// function getCasePermutations(input: string, index = 0, current = ''): string[] {
// 	if (index === input.length) {
// 		return [current]
// 	}

// 	const char = input[index]
// 	const nextPermutations = getCasePermutations(input, index + 1, current + char)

// 	if (char.toLowerCase() !== char.toUpperCase()) {
// 		const upperPermutations = getCasePermutations(input, index + 1, current + char.toUpperCase())
// 		return [...nextPermutations, ...upperPermutations]
// 	}

// 	return nextPermutations
// }
