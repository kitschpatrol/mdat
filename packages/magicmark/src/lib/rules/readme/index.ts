import { type RuleSet } from '../types'
import badges from './badges'
import code from './code'
import contributing from './contributing'
import footer from './footer'
import header from './header'
import license from './license'
import shortDescription from './short-description'
import tableOfContents from './table-of-contents'
import title from './title'

// A little redundant to have an expander name / preset in both
// the record keys and the `keyword` field in the expansion object,
// but this smooths over the kebab vs. camel case difference and arguably
// makes expansions more portable

export default {
	badges,
	code,
	contributing,
	footer,
	header,
	license,
	shortDescription,
	tableOfContents,
	title,
} satisfies RuleSet
