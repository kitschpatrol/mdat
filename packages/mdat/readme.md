<!--+ Warning: Content in HTML comment blocks generated by mdat on 2024-02-08 +-->

<!-- header -->

# mdat

[![NPM Package](https://img.shields.io/npm/v/mdat.svg)](https://npmjs.com/package/mdat)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

**CLI tool and library for using comments as content templates in Markdown files.**

<!-- /header -->

> \[!NOTE]\
> **Please see the [Mdat Monorepo readme](http://github.com/kitschpatrol/mdat) for additional context.**

<!-- table-of-contents -->

## Table of contents

- [Overview](#overview)
- [Getting started](#getting-started)
  - [Dependencies](#dependencies)
  - [Installation](#installation)
- [Usage](#usage)
  - [CLI](#cli)
  - [API](#api)
  - [Configuration](#configuration)
- [Creating custom rules](#creating-custom-rules)
- [Maintainers](#maintainers)
- [Acknowledgements](#acknowledgements)
- [Contributing](#contributing)
- [License](#license)

<!-- /table-of-contents -->

## Overview

This is a CLI tool and library implementing the Markdown Autophagic Template (mdat) system, which makes it easy to automate the replacement of placeholder comments in Markdown documents with dynamic content from a variety of sources. `mdat` also validates the structure and content of the Markdown document based on constraints specified in the expansion rules.

A trivial example...

Given placeholder comments in a Markdown file like this:

`some-file.md`

```md
<!-- title -->
```

Run your file through the tool:

```sh
mdat some-file.md
```

To turn it into:

`some-file.md`

```md
<!-- title -->

# mdat

<!-- /title -->
```

In this case, according to a set of rules defined in an external configuration file, `<!-- title -->` was replaced with date from `package.json`. The rule system behind these expansions is simple to define and readily extensible beyond the trivial example above.

The `mdat` CLI and library package sits between the lower-level \[`remark-mdat`] remark plugin, and the higher-level \[`mdat-readme`] package. `mdat` offers flexibility, but if you're just looking to expand some placeholders in your readme files then [`mdat-readme`](../mdat-readme) is the place to start.

## Getting started

### Dependencies

The `mdat` CLI tool requires Node 16+. The exported APIs for expanding Markdown text and documents are ESM-only and share the Node 16+ requirement. `mdat` is implemented in TypeScript and bundles a complete set of type definitions.

### Installation

Install locally to access the CLI commands in a single project or to import the provided APIs:

```sh
npm install remark-mdat
```

Or, install globally for access across your system:

```sh
npm install --global remark-mdat
```

## Usage

> \[!WARNING]\
> **The `mdat` CLI tool directly manipulates the contents of readme files, in close (and perhaps dangerous) proximity to your painstakingly crafted words.**
>
> Please make sure any text you care about is committed before running `mdat`, and never directly modify content inside of the comment expansion blocks.
>
> Set the `--meta` flag on the command to add a warning comment to the top of your file explaining the extra caution demanded around the volatile automated sections of your readme.md.

### CLI

<!-- cli-help -->

#### Command: `mdat`

Note: `expand` is the default and only command at the moment.

This section lists top-level commands for `mdat`.

If no command is provided, `mdat expand` is run by default.

Usage:

```txt
mdat [command] [options]
```

| Command  | Argument                | Description                                                         |
| -------- | ----------------------- | ------------------------------------------------------------------- |
| `expand` | `<files..>` `[options]` | Expand comment placeholders in Markdown files. _(Default command.)_ |

_See the sections below for more information on each subcommand._

#### Subcommand: `mdat expand`

Expand comment placeholders in Markdown files.

Usage:

```txt
mdat expand <files..> [options]
```

| Positional Argument | Description                                                                | Type     |
| ------------------- | -------------------------------------------------------------------------- | -------- |
| `files`             | Markdown file(s) with `mdat` placeholder comments to expand. _(Required.)_ | `string` |

| Option      | Alias | Description                                                                                                                                                                                                   | Type      | Default                                                                       |
| ----------- | ----- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------- | ----------------------------------------------------------------------------- |
| `--config`  |       | Path(s) to files containing mdat configs.                                                                                                                                                                     | `array`   | Configuration is loaded if found from the usual places, or defaults are used. |
| `--rules`   | `-r`  | Path(s) to files containing `mdat` comment expansion rules.                                                                                                                                                   | `array`   |                                                                               |
| `--output`  | `-o`  | Output file directory.                                                                                                                                                                                        | `string`  | Same directory as input file.                                                 |
| `--name`    | `-n`  | Output file name.                                                                                                                                                                                             | `string`  | Same name as input file. Overwrites the input file.                           |
| `--print`   |       | Print the expanded Markdown to stdout instead of saving to a file. Ignores `--output` and `--name` options.                                                                                                   | `boolean` | `false`                                                                       |
| `--prefix`  |       | Require a string prefix before all comments to be considered for expansion. Useful if you have a bunch of non-`mdat` comments in your Markdown file, or if you're willing to trade some verbosity for safety. | `string`  |                                                                               |
| `--meta`    | `-m`  | Embed an extra comment at the top of the generated Markdown noting the date of generation and warning editors that certain sections of the document have been generated dynamically.                          | `boolean` | `false`                                                                       |
| `--check`   | `-c`  | Check the input files for rule violations without expanding them. Identifies things like missing comment placeholders and incorrect placeholder ordering.                                                     | `boolean` | `false`                                                                       |
| `--verbose` |       | Enable verbose logging. All verbose logs and prefixed with their log level and are printed to stderr for ease of redirection.                                                                                 | `boolean` | `false`                                                                       |
| `--help`    | `-h`  | Show help                                                                                                                                                                                                     | `boolean` |                                                                               |
| `--version` | `-v`  | Show version number                                                                                                                                                                                           | `boolean` |                                                                               |

<!-- /cli-help -->

<!-- cli-help-note -->

_Meta note: The entire section above was generated automatically by the [`<!-- cli-help -->`](../mdat-readme/src/lib/rules/cli-help/index.ts) mdat expansion rule provided in `mdat-readme`. It dynamically parses the output from `mdat --help` into a markdown table, recursively calling `--help` on subcommands to build a Markdown representation of the help output._

<!-- /cli-help-note -->

#### Examples

##### Basic

Expand comments in a single Markdown file in-place:

```sh
mdat your-file.md
```

##### Multi-file

Expand comments in multiple Markdown files:

```sh
mdat *.md
```

##### Custom configuration

A number of option flags are exposed on the CLI. Any values set here will override both ambient configuration files and any configuration file referenced passed as options:

```sh
mdat  --prefix 'mm-
```

##### Custom configuration file

```sh
mdat  --config 'custom-config.ts"
```

##### Rule sets

```sh
mdat --rules 'rules.ts' 'more-rules.js' 'yet-more-rules.json'
```

### API

`mdat` exports a collection of functions to abstract the process of expanding placeholder comments into a single call. Type aliases are also provided.

Highlights include:

#### Expand String

```ts
function expandString(markdown: string, config?: ExpandConfig, rules?: ExpandRules): Promise<VFile>
```

Takes a string of Markdown and returns. Note that the returned object is a [VFile](https://github.com/vfile), which includes both the post-conversion Markdown content and additional metadata about the conversion.

To get the Markdown content, simply call `.toString()` on the returned VFile object.

#### Expand File

```ts
function expandFile(
  file: string,
  name?: string,
  output?: string,
  config?: ExpandConfig,
  rules?: ExpandRules,
): Promise<VFile>
```

Similar to `expandString()`, but takes a file path and handles setting an optional destination path and file name.

It's up to the caller to actually save the returned VFile object. The [to-vfile](https://www.npmjs.com/package/to-vfile) library can make this particularly painless:

```ts
import { expandFile } from 'mdat'
import { write } from 'to-vfile'

const file = await expandFiles(...)
await write(file)
```

#### Expand Files

```ts
function expandFiles(
  files: string[],
  name?: string,
  output?: string,
  config?: ExpandConfig,
  rules?: ExpandRules,
): Promise<VFile[]>
```

Like `expandFile()`, but accepts an array of inputs. If an output name is specified, the output files are suffixed with a number to prevent name collisions.

#### Load Config

```ts
function loadConfig<T extends Partial<Options>>(options?: {
  additionalConfig?: Array<T | string> | T | string | undefined
  additionalRules?: Array<Rules | string> | Rules | string | undefined
  configExtensionSchema?: z.ZodObject<any>
  searchFrom?: string
}): Promise<T>
```

This is provided for more advanced use cases. It assists in discovering and loading ambient configuration in your project (e.g. fields in your package.json, or dedicated `mdat` config files). It also dynamically loads, validates, and merges additional `mdat` configuration and rule files into a final `ExpandConfig` object ready to be passed into the [`remark-mdat`](../remark-mdat/readme.md) plugin or one of the API functions like `expandFile()`.

The function is generic to accommodate extensions to the configuration type in derivative projects. ([`mdat-readme`]() uses this approach.)

#### Examples

### Configuration

Expansion rules and certain aspects of global configuration are defined in configuration files which may be discovered automatically by `mdat`, or explicitly provided via command line options or library function arguments as shown above.

`mdat` implements configuration loading via [cosmiconfig](https://github.com/cosmiconfig/cosmiconfig), which means a variety of configuration [locations](https://github.com/cosmiconfig/cosmiconfig?tab=readme-ov-file#searchplaces) and file formats are supported. Configuration may be defined directly in your package.json, or in addition to stand-alone TypeScript files, JavaScript files, YAML, JSON, etc.

TypeScript or JavaScript with JSDoc annotations are recommended for the most flexibility and ease of implementing more advanced rules.

`mdat` also allows arbitrary JSON files to be loaded as rule sets, flattening them so any value may be accessed by using a dot-notation key path as a comment keyword.

#### Configuration file format

The `mdat` configuration file is a record object allowing you to customize aspects of the comment expansion process, and also optionally define expansion rules as well under the `rules` key:

```ts
type MdatConfig = {
  addMetaComment?: boolean
  closingPrefix?: string
  keywordPrefix?: string
  metaCommentIdentifier?: string
  rules?: Rules
}
```

A valid configuration file default-exports an object conforming to the above type.

The configuration file may be located in any location supported by [cosmicconfig](https://github.com/cosmiconfig/cosmiconfig?tab=readme-ov-file#searchplaces). I use a `mdat.config.ts` file in the root of my projects.

#### Rule file format

Rules may also be defined in separate files that default-export a record of rules. The record keys become the keywords used to reference a rule from your comments in Markdown.

```ts
type Rules = Record<string, Rule>

type Rule =
  | string
  | ((options: JsonValue, tree: Root) => Promise<string> | string)
  | Rule[]
  | {
      content: string | Rule[] | ((options: JsonValue, tree: Root) => Promise<string> | string)
      applicationOrder?: number | undefined
      order?: number | undefined
      required?: boolean | undefined
    }
```

This is a bit complex, but it's indented to make defining simple rules simple, while still accommodating more demanding use cases.

Some notes on the type:

- Simple rules can be defined directly on the key, either as strings to replace the comment placeholder, or as sync or async functions returning a string.

- If you need more advanced rules, or wish to define conditions for the validation process, you break the top-level keyword key's value out into an object, where a `content` key on the object is responsible for returning the replacement string, and additional fields are available to define validation constraints.

- Note that `content` can itself take an array of Rule objects, which is useful for creating "compound" rules that combine several others into a single comment keyword.

Since it's a record, multiple rules may be combined in single rules file.

If multiple configuration and rule files are loaded, they are merged. CLI options take precedence over ambient configuration discovered by cosmicconfig, and where multiple configuration or rule files are provided, with the "last" rule of a particular key takes precedence.

## Creating custom rules

See the [Examples section](../remark-mdat/readme.md#examples) of the `remark-mdat` readme, or take a look at the implementation of the [rules bundled with `mdat-readme`](../mdat-readme/src/lib/rules/) for more complex examples.

## Maintainers

[@kitschpatrol](https://github.com/kitschpatrol)

## Acknowledgements

Please see the [monorepo readme](../../readme.md#acknowledgments).

<!-- footer -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

## License

[MIT](license.txt) © Eric Mika

<!-- /footer -->
