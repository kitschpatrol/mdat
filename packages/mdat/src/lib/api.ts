import { type ConfigLoaded, type ConfigToLoad, type RulesToLoad, loadConfig } from './config'
import { getInputOutputPath, getInputOutputPaths } from './utilities'
import { type Root } from 'mdast'
import { remark } from 'remark'
import remarkGfm from 'remark-gfm'
import { type Options, mdatClean, mdatSplit, default as remarkMdat } from 'remark-mdat'
import { read } from 'to-vfile'
import { VFile } from 'vfile'

/**
 * Writing is the responsibility of the caller (e.g. via `await write(result)`)
 */
export async function expandFiles(
	files: string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile[]> {
	const resolvedConfig = await loadConfig({ additionalConfig: config, additionalRules: rules })
	// Store config for access in rules
	config = resolvedConfig

	// Does some validation and  adds  a number to the name if needed
	const inputOutputPaths = getInputOutputPaths(files, output, name, 'md')
	const results: VFile[] = []

	// We don't call expandFile so we can reuse the loadConfig output and processor
	const processor = getProcessor(resolvedConfig)
	for (const { input, name, output } of inputOutputPaths) {
		const inputFile = await read(input)
		const result = await processor.process(inputFile)
		result.dirname = output
		result.basename = name
		results.push(result)
	}

	return results
}

/**
 * Writing is the responsibility of the caller (e.g. via `await write(result)`)
 */
export async function expandFile(
	file: string,
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	const resolvedConfig = await loadConfig({ additionalConfig: config, additionalRules: rules })
	// Store config for access in rules
	config = resolvedConfig

	const inputOutputPath = getInputOutputPath(file, output, name, 'md')
	const inputFile = await read(inputOutputPath.input)
	const result = await getProcessor(resolvedConfig).process(inputFile)
	result.dirname = inputOutputPath.output
	result.basename = inputOutputPath.name
	return result
}

export async function expandString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	const resolvedConfig = await loadConfig({ additionalConfig: config, additionalRules: rules })
	// Store config for access in rules
	config = resolvedConfig

	return getProcessor(resolvedConfig).process(new VFile(markdown))
}

/**
 * Writing is the responsibility of the caller (e.g. via `await write(result)`)
 */
export async function cleanFiles(
	files: string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
): Promise<VFile[]> {
	const resolvedConfig = await loadConfig({
		additionalConfig: config,
	})
	// Store config for access in rules
	config = resolvedConfig

	// Does some validation and  adds  a number to the name if needed
	const inputOutputPaths = getInputOutputPaths(files, output, name, 'md')
	const results: VFile[] = []

	// We don't call expandFile so we can reuse the loadConfig output and processor
	const processor = getCleanProcessor(resolvedConfig)
	for (const { input, name, output } of inputOutputPaths) {
		const inputFile = await read(input)
		const result = await processor.process(inputFile)
		result.dirname = output
		result.basename = name
		results.push(result)
	}

	return results
}

/**
 * Writing is the responsibility of the caller (e.g. via `await write(result)`)
 */
export async function cleanFile(
	file: string,
	name?: string,
	output?: string,
	config?: ConfigToLoad,
): Promise<VFile> {
	const resolvedConfig = await loadConfig({
		additionalConfig: config,
	})
	// Store config for access in rules
	config = resolvedConfig

	const inputOutputPath = getInputOutputPath(file, output, name, 'md')
	const inputFile = await read(inputOutputPath.input)
	const result = await getCleanProcessor(resolvedConfig).process(inputFile)
	result.dirname = inputOutputPath.output
	result.basename = inputOutputPath.name
	return result
}

export async function cleanString(markdown: string, config: ConfigToLoad): Promise<VFile> {
	const resolvedConfig = await loadConfig({
		additionalConfig: config,
	})
	// Store config for access in rules
	config = resolvedConfig

	return getCleanProcessor(resolvedConfig).process(new VFile(markdown))
}

// Helpers

function getProcessor(options?: Options) {
	const processor = remark()
		// Hard-coding some style preferences here. Users who want different
		// settings can use remark-mdat to create their own processor.
		.use({
			settings: {
				bullet: '-',
				emphasis: '_',
			},
		})
		.use(remarkGfm)
		.use(remarkMdat, options ?? {})

	return processor
}

function getCleanProcessor(options: ConfigLoaded) {
	// MdatCleanOptions compatible object
	const { closingPrefix, keywordPrefix, metaCommentIdentifier } = options

	const processor = remark()
		// Hard-coding some style preferences here. Users who want different
		// settings can use remark-mdat to create their own processor.
		.use({
			settings: {
				bullet: '-',
				emphasis: '_',
			},
		})
		.use(remarkGfm)
		.use(
			() =>
				// eslint-disable-next-line unicorn/consistent-function-scoping
				function (tree: Root, file: VFile) {
					mdatSplit(tree, file)
					mdatClean(tree, file, { closingPrefix, keywordPrefix, metaCommentIdentifier })
				},
		)

	return processor
}
