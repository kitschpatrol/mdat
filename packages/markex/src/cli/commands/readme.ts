import { type ExpandAstOptions, expandString, presets } from '../../lib'
import log from '../../lib/log'
import { expandFile, getReadmePath } from '../helpers'
import fs from 'node:fs/promises'
import type { ArgumentsCamelCase, CommandModule, Options } from 'yargs'

export type CommandArguments = {
	meta: boolean
	prefix: string
	print: boolean
	verbose: boolean
} & Options

// Yargs command class approach via:
// https://gist.github.com/dsmrt/cf1812cb657738e83d76ba13ca28dbde
export class Readme<U extends CommandArguments>
	implements CommandModule<Record<string, unknown>, U>
{
	public command = 'readme'
	public describe = 'description to come'

	public handler = async (argv: ArgumentsCamelCase<U>) => {
		const { meta = true, prefix, print, verbose } = argv
		log.verbose = verbose

		const expandOptions: ExpandAstOptions = {
			expansionRules: presets.readme,
			keywordPrefix: prefix,
			meta,
		}

		// First, check for a readme in the package directory
		const readmePath = await getReadmePath()
		const readmeString = await fs.readFile(readmePath, 'utf8')
		const { expandedString, report } = await expandString(readmeString, expandOptions)

		if (!print) {
			await fs.writeFile(readmePath, expandedString)
		}

		log.info('[readme]', `Expanded:`)
		log.info('[readme]', `  From: ${readmePath}`)
		log.info('[readme]', `  To:   ${print ? 'stdout' : readmePath}`)
		log.info('[readme]', '  Replaced:')
		for (const [i, line] of report.entries()) {
			log.info('[readme]', `    ${i + 1}. ${line}`)
		}

		if (print) {
			process.stdout.write(readmeString ?? '')
		}
	}
}
