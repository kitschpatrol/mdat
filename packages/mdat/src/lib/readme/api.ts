import { cleanFile, expandFile, expandString } from '../api'
import {
	type Config,
	type ConfigLoaded,
	type ConfigToLoad,
	type RulesToLoad,
	loadConfig,
} from '../config'
import readmeRules from './rules'
import { type Simplify } from 'type-fest'
import { type VFile } from 'vfile'

// Readme and package files are mandatory for readme operations
type ReadmeConfigLoaded = Simplify<
	{
		packageFile: string
		readmeFile: string
	} & ConfigLoaded
>

export async function expandReadme(
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
		additionalRules: rules,
	})

	return expandFile(resolvedConfig.readmeFile, name, output, resolvedConfig)
}

export async function expandReadmeString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
		additionalRules: rules,
	})

	return expandString(markdown, resolvedConfig)
}

export async function cleanReadme(config?: ConfigToLoad): Promise<VFile> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
	})

	return cleanFile(resolvedConfig.readmeFile, undefined, undefined, resolvedConfig)
}

// Helpers

// Convenience loader to always include the default readme config
type LoadConfigOptions = Parameters<typeof loadConfig>[0]
export async function loadConfigReadme(options?: LoadConfigOptions): Promise<ReadmeConfigLoaded> {
	const defaultReadmeConfig: Config = {
		addMetaComment: true,
		rules: readmeRules,
	}

	const { additionalConfig = [], ...rest } = options ?? {}
	const additionalConfigArray = Array.isArray(additionalConfig)
		? additionalConfig
		: [additionalConfig]

	const result = await loadConfig({
		additionalConfig: [defaultReadmeConfig, ...additionalConfigArray],
		...rest,
	})

	// This should never happen because the defaults are set in loadConfigReadme
	if (result.packageFile === undefined || result.readmeFile === undefined) {
		throw new Error('Package and readme files are required in `mdat readme` config')
	}

	// Trust all optionals have been set
	return result as ReadmeConfigLoaded
}
