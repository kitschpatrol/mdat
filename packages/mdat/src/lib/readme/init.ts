/* eslint-disable perfectionist/sort-objects */
import { findPackage, findReadme } from '../utilities'
import { expandReadme } from './api'
import templates from './templates'
import { confirm, group, intro, note, outro, select } from '@clack/prompts'
import chalk from 'chalk'
import fs from 'node:fs/promises'
import path from 'node:path'
import { deepMergeDefined } from 'remark-mdat'
import { write } from 'to-vfile'

export type Symbolize<T extends Record<string, unknown>> = {
	[x in keyof T]: T[x] | symbol
}

type MdatReadmeInitOptions = {
	compound: boolean
	expand: boolean
	output: string
	overwrite: boolean
	template: string
}

async function getPaths(): Promise<{
	packageDirectory: string | undefined
	readmePath: string | undefined
}> {
	// Defaults, runs silently
	const readmePath = await findReadme()
	const packagePath = await findPackage()
	const packageDirectory =
		packagePath === undefined ? undefined : path.dirname(path.resolve(packagePath))

	return { packageDirectory, readmePath }
}

export async function initReadmeInteractive(): Promise<string> {
	const { packageDirectory, readmePath } = await getPaths()
	const destination = path.resolve(process.cwd())

	intro(`Running ${chalk.bold('mdat-readme init')} interactively`)

	const initConfig = await group<Symbolize<MdatReadmeInitOptions>>(
		{
			overwrite: async () =>
				readmePath === undefined
					? true
					: (await confirm({
								message: `Found an existing readme at "${chalk.blue(readmePath)}". Do you want to overwrite it?`,
								active: 'Overwrite',
								inactive: 'Exit',
						  }))
						? true
						: (() => {
								throw new Error(
									'`mdat-readme init` was cancelled to avoid an overwrite - no changes were made',
								)
							})(),

			output: async () =>
				packageDirectory !== undefined && packageDirectory !== destination
					? select({
							initialValue: packageDirectory,
							message:
								"There's a root package directory nearby, do you want to create the readme there instead of the current directory?",
							options: [
								{
									label: `Create in the current package root: "${chalk.blue(packageDirectory)}"`,
									value: packageDirectory,
								},
								{
									label: `Create in current working directory: "${chalk.blue(destination)}"`,
									value: destination,
								},
							],
						})
					: destination,

			template: async () =>
				select<TemplateOptions, string>({
					message: 'Which template would you like to use?',
					options: getTemplateOptions(),
				}),

			compound: async () =>
				select({
					message:
						'Do you want to use "compound comments" where possible, which combine several expansions into a single comment block?',
					options: [
						{
							label: `Yes: Combine things like ${chalk.green('<!-- title -->')} and ${chalk.green('<!-- badges -->')} in a single ${chalk.green('<!-- header -->')} comment.`,
							value: true,
						},
						{
							label: 'No:  Use individual `mdat` expansion comments for each section.',
							value: false,
						},
					],
				}),

			expand: async () =>
				packageDirectory === undefined
					? false
					: confirm({
							initialValue: true,
							message:
								'Do you want to run mdat-readme now to expand the template with content from your package.json?',
						}),
		},
		{
			// On Cancel callback that wraps the group
			// So if the user cancels one of the prompts in the group this function will be called
			onCancel() {
				throw new Error('`mdat-readme init` was cancelled - no changes were made')
			},
		},
	)

	// Do the deed
	const newReadmePath = await initReadme(initConfig)

	if (packageDirectory === undefined) {
		note(
			"No package.json was found. Once you've created one, you can run `mdat-readme` to expand the template with content from your package.json.",
		)
	}

	note(`Readme created: "${chalk.bold.blue(newReadmePath)}"`)

	outro('Done!')

	return newReadmePath
}

/**
 * @return Path to the created readme file.
 */
export async function initReadme(options?: Partial<MdatReadmeInitOptions>): Promise<string> {
	const { packageDirectory } = await getPaths()
	const resolvedOptions = deepMergeDefined(
		{
			compound: true,
			output: packageDirectory ?? process.cwd(),
			overwrite: true,
			expand: packageDirectory !== undefined,
			template: Object.keys(templates)[0],
		},
		options ?? {},
	) as Required<MdatReadmeInitOptions>

	// Save the template
	const templateString = getTemplateForConfig(resolvedOptions.template, resolvedOptions.compound)
	const readmePath = path.join(resolvedOptions.output, 'readme.md')
	await fs.writeFile(readmePath, templateString, 'utf8')

	// Run the expansion if requested
	// Maybe better to use execa?
	const results = await expandReadme(undefined, undefined, { readmeFile: readmePath })
	await write(results)

	return readmePath
}

function getTemplateForConfig(templateKey: string, compound: boolean): string {
	const templateObject = templates[templateKey as keyof typeof templates]
	const templateString = templateObject.content[compound ? 'compound' : 'explicit']

	if (templateString === undefined || templateString === '') {
		throw new Error(`No template found for "${templateKey}"`)
	}

	return templateString
}

type TemplateOptions = Array<{ hint?: string; label: string; value: string }>

function getTemplateOptions(): TemplateOptions {
	return Object.entries(templates).map(([key, value]) => ({
		label: key,
		hint: value.description,
		value: key,
	}))
}
