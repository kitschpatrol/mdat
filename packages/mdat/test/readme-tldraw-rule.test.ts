import { expandReadmeString } from '../src/lib/readme/api'
import os from 'node:os'
import { describe, expect, it } from 'vitest'

// These have long timeouts because going through puppeteer is slow

describe('tldraw image rule', () => {
	it('should expand tldraw images from local files', async () => {
		const markdown = `<!-- tldraw { src: "./test/assets/tldraw-sketch.tldr" } -->`
		const result = await expandReadmeString(markdown, {
			assetsPath: `${os.tmpdir()}/assets`,
		})
		expect(stripDynamic(result.toString())).toMatchSnapshot()
	}, 30_000)

	it('should expand tldraw images from remote URLs', async () => {
		const markdown = `<!-- tldraw { src: "https://www.tldraw.com/s/v2_c_JsxJk8dag6QsrqExukis4" } -->`
		const result = await expandReadmeString(markdown, {
			assetsPath: `${os.tmpdir()}/assets`,
		})
		expect(stripDynamic(result.toString())).toMatchSnapshot()
	}, 30_000)
})

// Replace matched dates with the placeholder text for stable snapshots
function stripDynamic(text: string): string {
	return text.replaceAll(/\s\d{4}-\d{2}-\d{2}\s/g, ' ****-**-** ')
}
