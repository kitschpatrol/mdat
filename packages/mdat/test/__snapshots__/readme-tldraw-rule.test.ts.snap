// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tldraw image rule > should expand tldraw images from local files 1`] = `
"<!--+ Warning: Content in HTML comment blocks generated by mdat on ****-**-** +-->

<!-- tldraw { src: "./test/assets/tldraw-sketch.tldr" } -->

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="../../../../../../var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/tldraw-sketch-132cbdb8-dark.svg">
  <source media="(prefers-color-scheme: light)" srcset="../../../../../../var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/tldraw-sketch-132cbdb8-light.svg">
  <img alt="tldraw diagram" src="../../../../../../var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/tldraw-sketch-132cbdb8-light.svg">
</picture>

<!-- /tldraw -->
"
`;

exports[`tldraw image rule > should expand tldraw images from remote URLs 1`] = `
"<!--+ Warning: Content in HTML comment blocks generated by mdat on ****-**-** +-->

<!-- tldraw { src: "https://www.tldraw.com/s/v2_c_JsxJk8dag6QsrqExukis4" } -->

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="../../../../../../var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/v2_c_JsxJk8dag6QsrqExukis4-c1827e44-dark.svg">
  <source media="(prefers-color-scheme: light)" srcset="../../../../../../var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/v2_c_JsxJk8dag6QsrqExukis4-c1827e44-light.svg">
  <img alt="tldraw diagram" src="../../../../../../var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/v2_c_JsxJk8dag6QsrqExukis4-c1827e44-light.svg">
</picture>

<!-- /tldraw -->
"
`;
